src/StreamJsonRpc/MessageHandlerBase.cs

```diff
@@ -189,7 +190,18 @@ public async ValueTask WriteAsync(JsonRpcMessage content, CancellationToken canc
                     {
                         cancellationToken.ThrowIfCancellationRequested();
                         await this.WriteCoreAsync(content, cancellationToken).ConfigureAwait(false);
-                        await this.FlushAsync(cancellationToken).ConfigureAwait(false);
+
+                        // When flushing, do NOT honor the caller's CancellationToken since the writing is done
+                        // and we must not throw OperationCanceledException back at them as if we hadn't transmitted it.
+                        // But *do* cancel flushing if we're being disposed.
+                        try
+                        {
+                            await this.FlushAsync(this.DisposalToken).ConfigureAwait(false);
+                        }
+                        catch (OperationCanceledException ex) when (ex.CancellationToken == this.DisposalToken)
```

Comment:
💭 Seems like the more general option here would be:

```suggestion
                        catch (OperationCanceledException ex) when (this.DisposalToken.IsCancellationRequested)
```